### ISSUE TITLE
Add support for mesh networking using webrtc datachannels

### ISSUE THREAD

**Issue by unicomp21** [kaspanet/rusty-kaspa#95]
This issue being opened to brainstorm possible use cases which "could" become possible if there was a layer bridging internet packets into the browser.

Link to starting discord convo:
https://discord.com/channels/599153230659846165/1001242675942531132/1059520395721584670

**Comment by unicomp21** [comment]
An example, tailscale does ip packet bridging to the browser using webrtc datachannels.

https://webvm.io/

**Comment by unicomp21** [comment]
spitball #1

Looking at the tcp listener concept, what use cases become possible, if instead we have a kaspa-listener running in a browser (ie server side) and others can connect to that endpoint w/ a tiny micro payment derived from pow in the client side browser instance?  And this network could be approached from the internet side as well, w/ real servers?

https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels

https://github.com/pion/webrtc/blob/master/e2e/test.html

**Comment by unicomp21** [comment]
Why are we using grpc streams instead of simple protobufs w/ an envelope protobuf message containing stream_correlation_uuid's ?

**Comment by unicomp21** [comment]
spitball /2

What if people could pay for ipfs pinning services by using kaspa nano-payments derived from work done in the browser node?

**Comment by unicomp21** [comment]
We might be able to get Brennan to add support for Kaska in the brave browser wallet?

**Comment by aspect** [comment]
Hello. I am one of the developers behind the KDX/PWA (kaspanet web wallet).  Unfortunately, web-based technologies are not anywhere as fast as they need to be for Kaspa, which is extremely high-performance as it needs to handle thousands of TPS.  WebRTC allows for high-speed data relay but then environments such as browsers, in the case of audio/video, rely on browser-supported A/V hardware acceleration, but themselves do not have performance capabilities to do custom data processing (at the level needed by Kaspa).  Even going with WebAssembly will not allow for the required performance.  Technology is not just there yet.

The original Golang implementation of kaspa is relying on gRPC.  In the Rust implementation, the RPC layer has already been reworked and is implemented using Rust Traits, making the RPC layer itself protocol/transport agnostic.  This allows us to plug in any RPC backend implementation with relative ease. gRPC implementation is currently being done to ensure backward compatibility but in rusty-kaspa this is not a limiting factor in any way.

Back to WebRTC - it is not possible to have Kaspa run in the browser due to performance requirements, however, this does not apply to the client-side transaction/wallet/utxoindex interface.  For that purpose, we are making sure that part of rusty-kaspa supports WASM as a build target, allowing core libraries to operate in the WASM (and thus the browser) environment.  

As rusty-kaspa allows for different RPC backends, I have been working on a side on WebSocket implementation (https://github.com/workflow-rs/workflow-websocket) that functions uniformly in the browser environment and in the native (bare-metal) environments.  The goal behind this is 1) to use binary data transmission to reduce overall data consumption 2) to allow the development of platform-neutral APIs that function uniformly, on bare metal, and in WASM environments.  WASM functions are easily exportable to JS allowing clients to simply specify the WebSocket endpoint and use functions directly.  This can then be exported to other language environments, if needed, or if desirable, other environments can use gRPC or JSON RPC (or whatever another custom RPC implementation is integrated).

To summarize, the current goal is to have a part of rusty-kaspa run in WASM environment and use WebSockets to communicate with full nodes (peers) running in native OS environments. Technically we can use JSON over WebSocket but currently, the aim is to use binary serialization as in my estimates it will reduce bandwidth consumption by half.

**Comment by unicomp21** [comment]
@aspect why websockets instead of rtcdatachannel?  realizing webassembly is less than ideal for mining, could there be any hope for mining nano-payments in the browser?  to pay for things like ipfs pinning services?

is there an argument for a mesh network composed of browsers nodes?  websockets require centralized servers, right?  the whole point of kaspa is decentralization?  rtc datachannels would allow p2p connections across browsers and servers w/o requiring centralized websocket servers?

**Comment by aspect** [comment]
I think what you may be referring to is more of a "browser-peer" as a "node" in kaspa terminology denotes a full-data-processing node, for which modern browsers have no sufficient processing capabilities.  Hence the idea is great but it is years ahead of what technology currently affords us. This has nothing to do with decentralization or anything like that. WebSocket is a way to bridge a browser environment with an application running directly on the CPU, nothing more.  WebSocket does not imply servers or is in any way related to centralization/decentralization topics.  KDX, for example, is a Web-enabled desktop application bundled with a full node using WebSockets to exchange information. Each user using it promotes node decentralization.

Potentially WebRTC can be used as an auxiliary relay of transaction data, but there is no reason to do that or at least to focus on that right now.  It is, in my opinion, an inappropriate hybridization of technology at this point in time. There are other elements at play here, such as a "full node" requiring a complex state-keeping, which is not feasible (or at least not performant) on the browser.  Kaspa node is a high-end data processor which, to achieve high performance/speed, relies on techniques such as complex optimizations on top of 64-bit multi-core CPU architectures, whereas a browser is a single-threaded 32-bit web-app platform that allows a maximum 2GB-4GB or low-performance RAM (whereas Kaspa's requires a minimum 8GB for DAG processing). etc. etc.

One of the goals behind Kaspa is to provide a fully decentralized high-performance / low-latency transactional network - any browser-related technologies other than UI/UX are not really feasible for anything like this.  A browser can not even render a video or play a sound, so it is not feasible for any type of high-performance data processing. (just re-iterating the point that browsers do not render videos, play sounds, or do any high-performance work - all that work is done by the application itself, not the website loaded into it).

Please continue on discord if you have any specific questions.

**Comment by unicomp21** [comment]
@aspect they asked me to create a github issue on discord for this discussion, so staying over here.  All the arguments posted above, does webgpu (currently in chrome origin trial) change any of this?  Or compute shaders based on webgl2 fragment shaders?

**Comment by unicomp21** [comment]
@michaelsutton

**Comment by unicomp21** [comment]
Mining pools, webrtc datachannel mesh networks, mining compute shaders based on webgpu or webgl.  Tensorflow has done a pretty good job proving out the idea of gpu compute in the web browser?

https://github.com/tensorflow/tfjs

Should we be be creating a bzminer equivalent for the web browser?
